### PR TITLE
presences: reconcile presence-init writes in bulk

### PR DESCRIPTION
**Pull request by ja0x**
Depends-On: https://github.com/wazo-platform/wazo-chatd/pull/201

## Summary

Reconcile presence-init writes in bulk, replacing the per-row create/update/delete in the presence initiator.

- Bulk DAO operations: chunked `bulk_insert`, set-based `UPDATE ... FROM VALUES`, and chunked `IN` deletes — routed through the DAO's injected session.
- Each resource is reconciled with a single diff against the cached state; only the rows that changed are written.
- Endpoints and channels are de-duplicated by name before insert.
- Total presence-initialization wall-time is logged at `INFO`.

## Tests

`test_presence_initialization` integration suite passes (10/10); unit tests and pre-commit (flake8/black/isort/mypy) clean.

---

> [!NOTE]
> **Medium Risk**
> Touches core presence DB reconciliation at startup (users, lines, sessions, tokens); behavior changes are covered by integration tests but incorrect bulk reconcile could skew presence state until the next init.
> 
> **Overview**
> **Replaces per-row presence initialization** with set-based reconciliation and chunked bulk SQL, aimed at faster startup and correct line/session/token sync.
> 
> Adds shared **`bulk_insert` / `bulk_update` / `bulk_delete`** helpers (1k-row batches) and extends tenant, user, line, session, refresh token, endpoint, and channel DAOs with **`create_all`**, targeted bulk deletes/updates, and lightweight cache queries (`list_uuids`, `list_dnd`, etc.). The **presence `Initiator`** diffs remote state against the DB and only writes deltas; **line moves** are handled by flushing expired line IDs then recreating missing ones; **DND** and **`mobile`** on sessions/tokens use bulk updates; **endpoints and channels** are keyed by name to drop duplicates before insert. **Pagination** now stops on empty pages and warns when totals drift; total init time is logged at **INFO**.
> 
> **Tests** cover bulk endpoint defaults, mobile flag refresh, dropped lines, and cross-user line reassignment, plus unit tests for dedup and pagination edge cases. **mypy** enforces complete defs on the new bulk DAO surface.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit def9cc3e642de6e1c9af30c65cbf1d509c0e9261. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: wazo_chatd/plugins/presences/initiator.py [hunk 1]
```diff
@@ -295,94 +289,86 @@ def _add_and_remove_lines(self, users):
         }
 
         lines_missing = lines - lines_cached
+        lines_expired = lines_cached - lines
+        existing_line_ids = {id_ for id_, _, _ in lines_cached}
+
         with session_scope():
+            user_uuids = self._dao.user.list_uuids()
```
Comment (DrPyser) [inline wazo_chatd/plugins/presences/initiator.py]:
```suggestion
            user_uuids = set(self._dao.user.list_uuids())
```
since user_uuids is used only for membership check, somewhat more performant to use a set for large number of users.

File: wazo_chatd/plugins/presences/initiator.py [hunk 2]
```diff
@@ -295,94 +289,86 @@ def _add_and_remove_lines(self, users):
...
+                if id_ in existing_line_ids or id_ in new_line_ids:
                     logger.warning(
                         'Line "%s" already created. Line multi-users not supported', id_
                     )
                     continue
-                line = Line(id=id_)
                 logger.debug('Create line "%s"', id_)
-                self._dao.user.add_line(user, line)
+                new_line_ids.add(id_)
+                new_lines.append(Line(id=id_, user_uuid=user_uuid))
```
Comment (DrPyser) [inline wazo_chatd/plugins/presences/initiator.py]: could collapse new_line_ids and new_lines into a single dict.

```suggestion
                new_lines[id_] = Line(id=id_, user_uuid=user_uuid)
```

Then conditional becomes
```
                if id_ in existing_line_ids or id_ in new_lines:
```

and declaration
```
         new_lines = {}
```

and dao call

```
            self._dao.line.create_all(list(new_lines.values()))

File: wazo_chatd/plugins/presences/initiator.py [hunk 3]
```diff
@@ -295,94 +289,86 @@ def _add_and_remove_lines(self, users):
...
+                endpoint_names.add(endpoint_name)
 
-                logger.debug('Create endpoint "%s"', endpoint_name)
-                self._dao.endpoint.create(Endpoint(name=endpoint_name))
+        with session_scope():
+            existing = self._dao.endpoint.list_names()
+            missing = []
+            for name in endpoint_names - existing:
+                logger.debug('Create endpoint "%s"', name)
+                missing.append(Endpoint(name=name))
```
Comment (DrPyser) [inline wazo_chatd/plugins/presences/initiator.py]:
```suggestion
            missing = [Endpoint(name=name) for name in endpoint_names - existing]
            logger.debug('Creating %d missing endpoints', len(missing))
```

if we really find value in enumerating endpoints, we can collapse into a single log
```
            logger.debug('missing endpoints: %s', missing)
```

File: wazo_chatd/plugins/presences/initiator.py [hunk 4]
```diff
@@ -295,94 +289,82 @@ def _add_and_remove_lines(self, users):
...
-                try:
-                    user = self._dao.user.get([tenant_uuid], user_uuid)
-                    line = self._dao.line.get(id_)
-                except UnknownUserException:
-                    logger.debug('Line "%s" already deleted', id_)
-                    continue
                 logger.debug('Delete line "%s"', id_)
-                self._dao.user.remove_session(user, line)
+                expired_ids.append(id_)
+            self._dao.line.delete_by_ids(expired_ids)
```
[the remark this answers is not in the document]
Comment (ja0x) [reply wazo_chatd/plugins/presences/initiator.py]: Already present in master; will be revised/fixed in another PR (to stay scoped)

[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_chatd/plugins/presences/initiator.py]: ref. https://app.notion.com/p/wazo/Line-creates-run-before-deletes-in-presence-init-bulk-operations-39daeaae6d2a81c6b0dbd5d4e5757e91

**Review by DrPyser — APPROVED** [review]
LGTM, tested on my stack. cuts down initialization time from 15s (with line presence view optimization) to 0.8s. 💪
